### PR TITLE
Add Gamma flag and Brightness control to node-red-node-pi-neopixel

### PR DESCRIPTION
**Pull request by cymplecy**
as per discussion https://groups.google.com/forum/#!topic/node-red/sLViJgOGFXQ

### CODE REVIEW

File: hardware/neopixel/neopixel.js [hunk 1]
```diff
@@ -35,6 +35,8 @@ module.exports = function(RED) {
         this.fgnd = n.fgnd || "128,128,128";
         this.mode = n.mode || "pcent";
         this.rgb = n.rgb || "rgb";
+        this.gamma = n.gamma;
```
Comment (dceejay) [inline hardware/neopixel/neopixel.js]: Doesn't this need to be defaulted also ?

File: hardware/neopixel/neopixel.js [hunk 2]
```diff
@@ -72,7 +77,7 @@ module.exports = function(RED) {
                                 node.bgnd = colors.getRGB(pay,node.rgb);
                                 pay = node.bgnd;
                             }
-                            else { node.warn("Invalid payload : "+pay); return; }
+                            else { node.warn("Invalid payload : "+pay); return; } //
```
Comment (dceejay) [inline hardware/neopixel/neopixel.js]: superfluous // ?

File: hardware/neopixel/neopixel.js [hunk 3]
```diff
@@ -102,7 +107,7 @@ module.exports = function(RED) {
             }
         }
 
-        node.child = spawn(piCommand, [node.pixels, node.wipe, node.mode]);
+        node.child = spawn(piCommand, [node.pixels, node.wipe, node.mode, node.gamma,node.brightness]);
```
Comment (dceejay) [inline hardware/neopixel/neopixel.js]: if gamma can be omitted  (not defaulted) - may a s well make it the last param instead... easier to skip

File: hardware/neopixel/package.json [hunk 4]
```diff
@@ -1,22 +1,25 @@
...
-    "name"          : "node-red-node-pi-neopixel",
-    "version"       : "0.0.16",
-    "description"   : "A Node-RED node to output to a neopixel (ws2812) string of LEDS from a Raspberry Pi.",
-    "dependencies"  : {
-    },
-    "repository" : {
-        "type":"git",
-        "url":"https://github.com/node-red/node-red-nodes/tree/master/hardware/neopixel"
+    "name": "node-red-node-pi-neopixel",
+    "version": "0.0.17.sw2",
```
Comment (dceejay) [inline hardware/neopixel/package.json]: .sw2 ?

File: hardware/neopixel/neopixel.html [hunk 5]
```diff
@@ -31,6 +31,16 @@
         <option value="grb">GRB</option>
     </select>
 </div>
+<div class="form-row">
```
Comment (dceejay) [inline hardware/neopixel/neopixel.html]: imho I'd prefer gamma after brightness if it's going to be default on and unlikely to be adjusted.

File: hardware/neopixel/neopixel.html [hunk 6]
```diff
@@ -31,6 +31,16 @@
...
+    <label for="node-input-gamma"> </label>
+    <label for="node-input-gamma" style="width:70%">
+    <input type="checkbox" id="node-input-gamma" style="display:inline-block; width:22px; vertical-align:baseline;"> Gamma corrected
+    </label>
+</div>
```
Comment (dceejay) [inline hardware/neopixel/neopixel.html]: Is there a few words that a layman can understand what this means ? (maybe not - but just asking)

File: hardware/neopixel/neopixel.js [hunk 7]
```diff
@@ -45,6 +47,9 @@ module.exports = function(RED) {
         var p4 = /^[0-9]+,[0-9]+,[0-9]+,[0-9]+,[0-9]+$/
 
         function inputlistener(msg) {
+            if (msg.hasOwnProperty("brightness")) {
+                node.child.stdin.write("brightness,"+msg.brightness.toString()+"\n");
```
Comment (dceejay) [inline hardware/neopixel/neopixel.js]: might want to check it really is 0 - 100 here...

File: hardware/neopixel/neopixel.js (hunk 1 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.js]: agreed

File: hardware/neopixel/neopixel.js (hunk 2 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.js]: I'll remove them

File: hardware/neopixel/neopixel.js (hunk 3 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.js]: AFAICT gamma cannot be omitted but I'll swap them around anyway

File: hardware/neopixel/package.json (hunk 4 above)
Comment (cymplecy) [reply hardware/neopixel/package.json]: I never know what do do re incrementing  someone elses code (as they might have uncommitted changes) but I'll just bump it up to plain 0.0.17

File: hardware/neopixel/neopixel.html (hunk 6 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.html]: I'll change it to "Apply gamma correction"

File: hardware/neopixel/neopixel.js (hunk 7 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.js]: I thought the earlier checks would make sure it was an number at this point but I'll re-look at it.  The python code does the full bounds checking/limiting at the moment

File: hardware/neopixel/neopixel.html (hunk 5 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.html]: I'll swap them around

File: hardware/neopixel/neopixel.js (hunk 7 above)
Comment (dceejay) [reply hardware/neopixel/neopixel.js]: maybe it does - but stops anyone doing something dumb and passing some large object over by mistake.

File: hardware/neopixel/neopixel.html [hunk 8]
```diff
@@ -53,6 +63,8 @@
     <!-- <p>The whole strip is set by <code>msg.payload</code> with a CSV string <i>r,g,b</i> -->
     <p>A range of pixels from <i>x</i> to <i>y</i> can be set by <code>msg.payload</code>
     with a CSV string <i>x,y,r,g,b</i>
+    <p>By default ,gamma correction is enabled but it can disabled which can be useful for working with low brightness levels</p>
```
Comment (dceejay) [inline hardware/neopixel/neopixel.html]: usually is  comma then space...   default, gamma

File: hardware/neopixel/neopixel.js (hunk 7 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.js]: Done the bounds checking at same time as bounds checking done for wipe time

File: hardware/neopixel/neopixel.js (hunk 1 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.js]: done

File: hardware/neopixel/neopixel.js (hunk 2 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.js]: removed

File: hardware/neopixel/neopixel.js (hunk 3 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.js]: order swapped

File: hardware/neopixel/package.json (hunk 4 above)
Comment (cymplecy) [reply hardware/neopixel/package.json]: to 0.0.17

File: hardware/neopixel/neopixel.html (hunk 5 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.html]: swapped

File: hardware/neopixel/neopixel.html (hunk 6 above)
Comment (cymplecy) [reply hardware/neopixel/neopixel.html]: changed to "Apply gamma correction"

**Comment by dceejay** [comment]
great  many thanks !

**Pull request merged**